### ISSUE TITLE
Rename happens in wrong location when build has failed

### ISSUE THREAD

**Issue by Kingdutch** [rescript-lang/rescript-vscode#417]
It looks like the rename happens based on the last known location for a token. This means that if you create a syntax error in your script and try to rename something, the location of the tokens may be outdated and the new name is inserted in the wrong position.

Video is attached that shows this happening. The syntax error in this case is an unbalanced piece of JSX.

https://user-images.githubusercontent.com/327697/168043168-d86d788d-e31a-4f8f-900d-d4b248a0e040.mov

**Comment by cristianoc** [comment]
Good point. I think rename should not happen when the file is unsaved.
@zth what do you think?

The server already keeps track of the current content of the file. Perhaps the server should also keep track of whether the file is saved.

**Comment by zth** [comment]
Yeah, maybe we should prompt the user that the file isn't saved, and then ask them to save it and try again?

**Comment by Kingdutch** [comment]
Shouldn't the extension be using the code that's in the editor's memory and not what's on disk? Or is there something that prevents that?

There may be reasons why the user does not want to save.

**Comment by cristianoc** [comment]
> Shouldn't the extension be using the code that's in the editor's memory and not what's on disk? Or is there something that prevents that?
> 
> 
> 
> There may be reasons why the user does not want to save.

Renaming with unsaved files is asking for trouble.

**Comment by cristianoc** [comment]
Looks like the server can receive notifications when a file is saved, but a way to "un-save" a file is to use undo, and there's no special notification that helps figuring out the file is not unsaved anymore as far as I can see.

So one might have to take the current file and compare it with the one on disk. Unclear whether this is best done on the server TS code or on the binary commands.

**Comment by zth** [comment]
I guess we'll need to check every file that might be affected by rename, and bail the rename if either of those is unsaved. I think we'll need to mull on this a bit more before going forward.
